external-crates/move/tools/move-package/tests/test_sources/external_overlap_fail_symmetric/Move.resolved

```diff
@@ -1,6 +1,6 @@
-Failed to resolve dependencies for package 'Root': Adding dependencies from ../resolvers/successful_dep.sh for dependency 'A' in 'Root': Conflicting dependencies found for 'ADep' during external resolution by '../resolvers/successful_dep.sh':
+When resolving external dependencies for package Root,conflicting dependencies found for 'ADep' during resolution by '../resolvers/successful_dep.sh':
 
-External dependencies not found:
-	{ name = "ADepDep" }
-New external dependencies:
-	{ name = "ADepOther" }
+Not found:
+	dependency { name = "ADepDep" } for package { local = "deps_only/ADepDep" }
```

Comment:
This message is admittedly not the best, and as much as I tried, I could not come up with anything better. Previously, we had this message only apply to problems due to external resolution, in which case the original formulation made sense - when we checked if dependencies of external package and internal package were the same, we could have either new external deps added (not present in the internal package) or missing (not found) external dependencies (those actually present in the internal package).

Now we use the same algorithm for comparing internal dependencies as well. It's a similar concept to the one in the previous version of the algorithm - we need to make sure that what was inserted into the main graph from one sub-graph matches what is being inserted from another sub-graph. This is needed for cases such as in the "diamond_problem_dep_incorrect_override_empty" test:

```
#                 +----+
#           +---->| B  |----+
#           |     +----+    |
#           |               |
# +----+    |               |     +----+ v1      +----+
# |Root|----+               +---->| C  |-------->| D  |
# +----+    |               |     +----+    +--->+----+
#           |               |               |
#           |     +----+----+      ov(2)    |
#           +---->| A  |--------------------+
#                 +----+
```
In this example we have a sub-graph for `A` (containing `C` and `D`) and a subgraph for `B` (also containing `C` and `D`). When processing sub-graph for `A` the dependency override in for `D` is valid and is applied, resulting in package `D` being at version `2.0.0`. In sub-graph for `B`, though, `D` is at version `1.0.0` as there is not override and the version that `C` demands wins.

After we merged sub-graph for `A` into root, we already have `C` and `D` in the main graph, and when we try merging `B` we see that `C` is already present. If we simply stop there, then a globally incorrect override (in `A`) is remains applied and generated main graph is incorrect. Instead, we need to make sure that `C` (now) in the main graph and `C` in the sub-graph being merged (in the sub-graph for `B`) are the same, and not only by comparing dependency edges in the graphs (which are actually the same in this case) but also by comparing the target packages.

Sorry for the overlong explanation, but I wanted to make sure that there is enough context for how (at least to me) the error message here was non-obvious...

Reply:
Thanks for clarifying @awelc -- the issue makes sense, but the fix (for me) seems to be the change in architecture that we talked about a bit before where rather than taking each graph and merging it into a main graph, we merge in bulk, I will outline it below, let me know if it makes sense, and whether or not I've missed a complication, but this I think is the key to simplifying the merging process:

- Get the `DependencyGraph`s for each immediate dependency (one for `A` and one for `B`).  Once we have these, whether a dependency in these graphs is an override or not is irrelevant.
- In each of these graphs, prune out the packages that are only accessible via a package that is overridden in the root, including the overridden package (in this example there are none of these).
  - (I would use a standard "mark and sweep" kind of algorithm for this).
  - This deals with the complication of a conflict that gets eliminated by an override of a dependent package.
- Take the union of all remaining packages in graphs from `A` and `B`, and merge them, elementwise :
   - `A`, `B` and `C` are trivially merged because they only appear in one of the graphs.
   - `D` presents as a conflict because it is required at version 1 in `B`'s graph and version 2 in `A`'s graph.
   - (The fix would be to add an override in `Root`'s manifest for `D`, which we would be aware of when we are doing the merge -- if we saw such an override we would pick that instead of either of `A`'s or `B`'s version of `D`).
- The error we give in this case is that `A` requires `D` at version `2` and `B` requires `D` at version `1`, and we can also offer the path from `A` to `D` in `A`'s graph and likewise in `B`'s graph for `B` -- by processing transitive dependencies in this way, we should have all the data at hand to provide this error.
- There is another kind of conflict that you can get, where packages mismatch because they depend on different dependencies but are otherwise the same -- the context for this kind of error is the same (i.e. share the paths from the dependency package to the transitive dependency), but the messaging would be different (this is related to the comment I made elsewhere around the asymmetry between internal and external resolution -- i.e. it shouldn't be asymmetric).

Reply:
I initially thought that "bulk merging" would mostly simplify override handling (which it already did) and the actual method of merging the graphs would not matter that much but this makes sense! I I will give it a shot to see if something unexpected does not creep up.